### PR TITLE
Allow tweaking the time between empty blocks

### PR DESCRIPTION
**Pull request by kim**
1 second is often too noisy. Cannot be set for `Production`, obviously.

### CODE REVIEW

File: src/Oscoin/Node.hs [hunk 1]
```diff
@@ -127,23 +130,39 @@ miner
...
                 liftIO $ tele (BlockMinedEvent bhash)
                 lift (tryBroadcast b) >>= liftIO . tele . \case
                     Left  e  -> BlockBroadcastFailedEvent bhash (toException e)
                     Right () -> BlockBroadcastEvent bhash
+
+        done  <- liftIO $ getTime Monotonic
+        let
+            !delta     = fromInteger . toNanoSecs $ diffTimeSpec start done
+            isProd     = Production == env
+            shouldWait = nTxs == 0 && not isProd && delta < emptyBlockTime
```
Comment (kim) [inline src/Oscoin/Node.hs]: Maybe it could apply regardless of the number of txs for this block, idk

**Comment by cloudhead** [comment]
It would make more sense if this was adjusting the `blockTime` (https://github.com/oscoin/oscoin/blob/master/src/Oscoin/Consensus/Nakamoto.hs#L39) as a consensus parameter, since that is exactly the purpose of it - otherwise we have two notions of "block time"

**Comment by cloudhead** [comment]
Alternatively, you can just update `blockTime` to `1 * minute` :joy: since that is the actual block time we're aiming for. This will work so long as our difficulty adjustment code works :crossed_fingers:

**Comment by kim** [comment]
If that only had the same effect on avoiding useless CPU cycles lol. Open to suggestions for alternative naming to avoid clashing with jargon.

**Comment by cloudhead** [comment]
Haha true. Maybe a better approach then would be to have a miner function which has a configurable delay in it? I'm reluctant to have debug/testing logic such as this in the main code path :thinking: - it would be nice if we could rather do something like `nakamotoConsensus { miner = delayedNakamotoMiner (1000 * seconds) }`, and have that isolated there.

**Comment by kim** [comment]
Hm yeah, we can move this to the `Lenient` module

File: src/Oscoin/Node/Options.hs [hunk 2]
```diff
@@ -58,6 +60,14 @@ nodeOptionsParser cps = Options
...
        <> showDefault
         )
+    <*> option (map (* 1000000000) auto)
+        ( long "empty-block-time"
+       <> help "Wait approx. this many seconds between empty blocks. Has no \
+               \effect when running with production settings."
+       <> metavar "SECONDS"
+       <> value 1
+       <> showDefault
+        )
```
Comment (cloudhead) [inline src/Oscoin/Node/Options.hs]: Should we prefix this option to communicate that it affects the lenient consensus? Eg. `lenient.empty-block-time` or something?

Comment (cloudhead) [reply src/Oscoin/Node/Options.hs]: We can also punt on this and revisit our whole cli options scheme at some point in the future :man_shrugging:

**Comment by kim** [comment]
Ait, it seems to me that consensus-affecting parameters are a bit of a mess at the moment, and it is easy to adjust one but miss out on others (and end up with non-meaningful permutations). I'm leaving this to someone else to consolidate in the future, including the decision if it should be possible to adjust the configuration from the outside (probably not).

Anyways, I'm amending this as follows:

* The target Nakamoto `blockTime` is set to 10 minutes (which seems to be the expectation elsewhere)
* `--no-empty-blocks` is removed, because correctness
* When running in `Development`, the (actual) block time can be tweaked using `--block-time-lower`, which imposes a lower bound. Ie. if there are no transactions to go into the block, the (lenient) miner waits for `min blockTimeLower blockTime` before actually mining the block. This avoids busy looping, but also control over how long to wait for a new block during development. Note that the main motivation here is the `devnet`, however, which will be idle most of the time, but shouldn't waste CPU (which is actually something cloud providers punish you for).

You may cast your vote for alternative names of the parameter here:

* `--hodgepodge-higgeldy-piggeldy`
* `--wait-mine-empty-block-seconds`
* `--lenient-empty-block-wait-seconds`
* `--seconds-before-mining-empty-block-in-development-environment`
* `--wat-gimme-a-break`

**Comment by xla** [comment]
`--lemme-save-some-trees`

`--abiding-cloud-citizen`

File: src/Oscoin/Consensus/Nakamoto.hs [hunk 3]
```diff
@@ -37,7 +37,7 @@ import           Database.SQLite.Simple.ToField (ToField)
 
 -- | Target block time.
 blockTime :: Duration
-blockTime = 1 * seconds
+blockTime = 10 * minutes
```
Comment (cloudhead) [inline src/Oscoin/Consensus/Nakamoto.hs]: The target for the mainnet will likely be `1 * minutes` - that's what we've been modeling our supply curves from, with `60 osc` per block. This gives you a nice and intuitive `1 osc / second` emission rate.

**Comment by cloudhead** [comment]
This does simplify things! I'm still in favor of `--lenient.*` as a namespace, but why not just have `--lenient.block-time`, and have it just be a way to overwrite the block-time? Ie. instead of using `min`, just have it be an override?

**Comment by kim** [comment]
> but why not just have --lenient.block-time, and have it just be a way to overwrite the block-time?

Here's why:

> I'm leaving this to someone else to consolidate in the future, including the decision if it should be possible to adjust the configuration from the outside (probably not).

**Pull request merged**